src/sync/impl/sync_metadata.cpp

```diff
@@ -189,110 +212,159 @@ bool SyncMetadataManager::delete_metadata_action(const std::string& original_nam
+                                                                                bool make_if_absent) const
 {
     // Open the Realm.
-    m_realm = Realm::get_shared_realm(manager.get_configuration());
+    auto realm = Realm::get_shared_realm(m_metadata_config);
+    auto& schema = m_user_schema;
 
     // Retrieve or create the row for this object.
-    TableRef table = ObjectStore::table_for_object_type(m_realm->read_group(), c_sync_userMetadata);
-    size_t row_idx = table->find_first_string(m_schema.idx_identity, identity);
-    if (row_idx == not_found) {
+    TableRef table = ObjectStore::table_for_object_type(realm->read_group(), c_sync_userMetadata);
+    Query query = table->where().equal(schema.idx_identity, identity).equal(schema.idx_auth_server_url, url);
+    Results results(realm, std::move(query));
+    REALM_ASSERT_DEBUG(results.size() < 2);
+    auto maybe_row_expr = results.first();
+    size_t row_idx = not_found;
+
+    if (!maybe_row_expr) {
         if (!make_if_absent) {
-            m_invalid = true;
-            m_realm = nullptr;
-            return;
+            return none;
         }
-        m_realm->begin_transaction();
-        row_idx = table->find_first_string(m_schema.idx_identity, identity);
-        if (row_idx == not_found) {
+        realm->begin_transaction();
+        // Check the results again.
+        auto maybe_row_expr = results.first();
+        if (!maybe_row_expr) {
             row_idx = table->add_empty_row();
-            table->set_string(m_schema.idx_identity, row_idx, identity);
-            table->set_bool(m_schema.idx_user_is_admin, row_idx, false);
-            m_realm->commit_transaction();
+            std::string uuid = util::uuid_string();
+            table->set_string(schema.idx_identity, row_idx, identity);
+            table->set_string(schema.idx_auth_server_url, row_idx, url);
+            table->set_string(schema.idx_local_uuid, row_idx, uuid);
+            table->set_bool(schema.idx_user_is_admin, row_idx, false);
+            table->set_bool(schema.idx_marked_for_removal, row_idx, false);
+            realm->commit_transaction();
         } else {
             // Someone beat us to adding this user.
-            m_realm->cancel_transaction();
+            row_idx = maybe_row_expr->get_index();
+            realm->cancel_transaction();
         }
-    }
-    m_row = table->get(row_idx);
-    if (make_if_absent) {
-        // User existed in the table, but had been marked for deletion. Unmark it.
-        m_realm->begin_transaction();
-        table->set_bool(m_schema.idx_marked_for_removal, row_idx, false);
-        m_realm->commit_transaction();
-        m_invalid = false;
     } else {
-        m_invalid = m_row.get_bool(m_schema.idx_marked_for_removal);
+        row_idx = maybe_row_expr->get_index();
+    }
+    REALM_ASSERT(row_idx != not_found);
+    auto row = table->get(row_idx);
+    if (make_if_absent || !row.get_bool(schema.idx_marked_for_removal)) {
```

Comment:
I'm having a hard time understanding this logic. It seems to read like this: if `make_if_absent` is true, or the user is not marked for removal… then ensure the user is not marked for removal, and return the the user. Is that correct? What's the purpose of the unconditional write transaction?